Math/NumberTheory/Primes/Factorisation/QuadraticSieve.hs

```diff
@@ -14,173 +14,222 @@ import Data.Semigroup
-findSmoothNumbers startingPoint = V.imapMaybe selectSmooth
+findLogSmoothNumbers :: Int -> Int -> Integer -> Integer -> V.Vector (Integer, Int) -> V.Vector (Integer, I.IntMap Int)
+findLogSmoothNumbers t m a b = V.imapMaybe selectSmooth
   where
-    selectSmooth index (residue, factorisation)
-      | residue == 1 = Just (intToInteger index + startingPoint, factorisation)
-      | otherwise    = Nothing
-
--- Find all primes, which appear only once in the input list.
-appearsOnlyOnce :: [PS.PrimeIntSet] -> PS.PrimeIntSet
-appearsOnlyOnce = fst . L.foldl' go (mempty, mempty)
-  where
-    go (onlyOnce, atLeastOnce) x =
-      ((onlyOnce PS.\\ PS.unPrimeIntSet x) <> (x PS.\\ PS.unPrimeIntSet atLeastOnce), atLeastOnce <> x)
+    -- Adjust fudge factor to taste
+    selectSmooth index (value, logResidue) = case logResidue < 18 of
+      True
+        -- Cannot use factorisation instead of listFactorisation since its rightmost value need not be an int.
+        | null listFactorisation || (fst . last) listFactorisation <= intToInteger t -> Just (a * intToInteger (index - m) + b, factorisation)
+        | otherwise                                                                  -> Nothing
+      False                                                                          -> Nothing
+      where
+        factorisation = if value < 0 then I.insert (-1) 1 preFac else preFac
+        preFac = I.fromAscList $ map (\(p, e) -> (integerToInt p, wordToInt e)) listFactorisation
+        -- Maybe there is a better way to use @trialDivision@
+        listFactorisation = trialDivisionTo (intToInteger t) value
 
 -- Removes all columns of the matrix which contain primes appearing only once.
 -- These columns cannot be part of the solution.
-removeRows :: [(Integer, SignedPrimeIntSet)] -> [(Integer, SignedPrimeIntSet)]
+-- This is a very lazy way to do it. There is probably a better one.
+removeRows :: [(Integer, I.IntMap Int)] -> [(Integer, I.IntMap Int)]
 removeRows indexedFactorisations
   | onlyOnce == mempty = indexedFactorisations
-  | otherwise          = removeRows $ filter (\(_, SignedPrimeIntSet _ xs) -> PS.disjoint xs onlyOnce) indexedFactorisations
+  | otherwise          = removeRows $ filter (\(_, im) -> S.null (S.intersection (convertToSet im) onlyOnce)) indexedFactorisations
   where
-    onlyOnce = appearsOnlyOnce $ map (primeSet . snd) indexedFactorisations
+    onlyOnce = appearsOnlyOnce $ map (convertToSet . snd) indexedFactorisations
+
+-- Find all primes, which appear only once in the input list.
+appearsOnlyOnce :: [S.IntSet] -> S.IntSet
+appearsOnlyOnce = fst . L.foldl' go (mempty, mempty)
+  where
+    go (onlyOnce, atLeastOnce) x =
+      ((onlyOnce S.\\ x) <> (x S.\\ atLeastOnce), atLeastOnce <> x)
+
+convertToSet :: I.IntMap Int -> S.IntSet
+convertToSet = I.foldrWithKey (\key pow set -> if odd pow then key `S.insert` set else set) mempty
```

Comment:
I'd rather filter `IntMap` by condition and used `I.keysSet` next.

Reply:
It's much better